apps/emqx/src/emqx_ocsp_cache.erl

```diff
@@ -0,0 +1,522 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2022-2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% @doc EMQX OCSP cache.
+%%--------------------------------------------------------------------
+
+-module(emqx_ocsp_cache).
+
+-include("logger.hrl").
+-include_lib("public_key/include/public_key.hrl").
+-include_lib("ssl/src/ssl_handshake.hrl").
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+-behaviour(gen_server).
+
+-export([
+    start_link/0,
+    sni_fun/2,
+    fetch_response/1,
+    register_listener/2,
+    inject_sni_fun/2
+]).
+
+%% gen_server API
+-export([
+    init/1,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2,
+    code_change/3
+]).
+
+%% internal export; only for mocking in tests
+-export([http_get/2]).
+
+-define(CACHE_TAB, ?MODULE).
+-define(CALL_TIMEOUT, 20_000).
+-define(RETRY_TIMEOUT, 5_000).
+-define(REFRESH_TIMER(LID), {refresh_timer, LID}).
+-ifdef(TEST).
+-define(MIN_REFRESH_INTERVAL, timer:seconds(5)).
+-else.
+-define(MIN_REFRESH_INTERVAL, timer:minutes(1)).
+-endif.
+
+-define(WITH_LISTENER_CONFIG(ListenerID, ConfPath, Pattern, ErrorResp, Action),
```

Comment:
I was using a macro because `Pattern` is pattern to match on when fetching the config; if the pattern fails, then `ErrorResp` is returned.

Maybe we could do the same with a normal function and then catch a `function_clause` error.  Do you think that would be better?

Reply:
Actually, since it's used in only 2 places now, it's not so bad to repeat.  I'll change it to a simple function, it's indeed better.  Thanks!

Reply:
although the code looks cleaner, It is really hard to trace or debug macros in erlang, I am just asking if it is possible to avoid that.
If you think macro helps in your case, just try to make it smaller.